mempool/mempool.go

```diff
@@ -525,12 +528,122 @@ func (mp *TxPool) IsOrphanInPool(hash *chainhash.Hash) bool {
 	return inPool
 }
 
+// isTransactionStaged determines if the transaction exists in the
+// stage pool.
+func (mp *TxPool) isTransactionStaged(hash *chainhash.Hash) bool {
+	_, exists := mp.staged[*hash]
+	return exists
+}
+
+// stageTransaction creates an entry for the provided
+// transaction in the stage pool.
+func (mp *TxPool) stageTransaction(tx *dcrutil.Tx) {
+	mp.staged[*tx.Hash()] = tx
+	for _, txIn := range tx.MsgTx().TxIn {
+		mp.stagedOutpoints[txIn.PreviousOutPoint] = tx
+	}
+}
+
+// removeStagedTransaction removes the provided transaction
+// from the stage pool.
+func (mp *TxPool) removeStagedTransaction(stagedTx *dcrutil.Tx) {
+	delete(mp.staged, *stagedTx.Hash())
+	for _, txIn := range stagedTx.MsgTx().TxIn {
+		delete(mp.stagedOutpoints, txIn.PreviousOutPoint)
+	}
+}
+
+// hasMempoolInput returns true if the provided transaction
+// has an input in the main pool
+func (mp *TxPool) hasMempoolInput(tx *dcrutil.Tx) bool {
+	for _, txIn := range tx.MsgTx().TxIn {
+		if mp.isTransactionInPool(&txIn.PreviousOutPoint.Hash) {
+			return true
+		}
+	}
+
+	return false
+}
+
+// fetchRedeemers returns all transactions that reference an outpoint for
+// the given transaction `tx`
```

Comment:
Missing full stop at the end.

Also, I'd document that this only works for regular transactions.